src/t_hash.c

```diff
@@ -297,25 +300,399 @@ static void hashDictWithExpireOnRelease(dict *d) {
+            break;
+
+        expired++;
+        fptr = lpNext(lpt->lp, fptr);
+    }
+
+    return expired;
+}
+
+/* Returns the expiration time of the item with the nearest expiration. */
+static uint64_t listpackExGetMinExpire(robj *o) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_EX);
+
+    long long expireAt;
+    unsigned char *fptr, *s;
+    listpackEx *lpt = o->ptr;
+
+    /* As fields are ordered by expire time, first field will have the smallest
+     * expiry time. Third element is the expiry time of the first field */
+    fptr = lpSeek(lpt->lp, 2);
+    if (fptr != NULL) {
+        s = lpGetValue(fptr, NULL, &expireAt);
+        serverAssert(!s);
+
+        /* Check if this is a non-volatile field. */
+        if (expireAt != HASH_LP_NO_TTL)
+            return expireAt;
+    }
+
+    return EB_EXPIRE_TIME_INVALID;
+}
+
+/* Walk over fields and delete the expired ones. */
+static void listpackExExpire(robj *o, ExpireInfo *info) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_EX);
+    uint64_t min = EB_EXPIRE_TIME_INVALID;
+    unsigned char *ptr, *field, *s;
+    listpackEx *lpt = o->ptr;
+
+    ptr = lpFirst(lpt->lp);
+    while (ptr != NULL && (info->itemsExpired < info->maxToExpire)) {
+        long long val;
+
+        field = ptr;
+        ptr = lpNext(lpt->lp, ptr);
+        serverAssert(ptr);
+        ptr = lpNext(lpt->lp, ptr);
+        serverAssert(ptr);
+
+        s = lpGetValue(ptr, NULL, &val);
+        serverAssert(!s);
+
+        /* Fields are ordered by expiry time. If we reached to a non-expired
+         * field or a non-volatile field, we know rest is not yet expired. */
+        if (val == HASH_LP_NO_TTL || (uint64_t) val > info->now)
+            break;
+
+        server.stat_expired_hash_fields++;
+        lpt->lp = lpDeleteRangeWithEntry(lpt->lp, &field, 3);
+        ptr = field;
+        info->itemsExpired++;
+    }
+
+    min = hashTypeGetNextTimeToExpire(o);
+    info->nextExpireTime = (min != EB_EXPIRE_TIME_INVALID) ? min : 0;
+}
+
+/* Remove TTL from the field. */
+static void listpackExPersist(robj *o, sds field, unsigned char *fptr,
+                              unsigned char *vptr)
+{
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_EX);
+
+    unsigned char tmp[512];
+    unsigned int slen;
+    long long val;
+    unsigned char *s;
+    sds p = NULL;
+    listpackEx *lpt = o->ptr;
+
+    /* To persist a field, we have to delete it first and append to the end as
+     * we want to maintain order by expiry time. Before deleting it, copy the
+     * value if it is stored as string. */
+    s = lpGetValue(vptr, &slen, &val);
+    if (s) {
+        /* Normally, item length in the listpack is limited by
+         * 'hash-max-listpack-value' config. It is unlikely, but it might be
+         * larger than sizeof(tmp). */
+        if (slen > sizeof(tmp))
+            p = sdsnewlen(s, slen);
+        else
+            memcpy(tmp, s, slen);
+    }
```

Comment:
maybe we can use `memmove` to move these three entries in the future improvement.

Reply:
I was thinking about lpBatchInsert() which can be used appending to the tail as well but probably moving is good enough.